### PR TITLE
Put Sequencer checkpoints into queue so they can be replicated

### PR DESCRIPTION
**Pull request by antonyshchenko**

### CODE REVIEW

File: sequencer/src/main/kotlin/co/chainring/sequencer/core/SequencerState.kt [hunk 1]
```diff
@@ -70,88 +88,56 @@ data class SequencerState(
...
+                            withdrawalFee {
+                                this.asset = it.key.value
+                                this.value = it.value.toIntegerValue()
+                            }.toByteArray()
+                        },
+                    )
+                    .write("balances")
+                    .bytes(getBalancesCheckpoint().toByteArray())
+                    .write("markets")
+                    .sequence(
```
Comment (bholzman) [inline sequencer/src/main/kotlin/co/chainring/sequencer/core/SequencerState.kt:133]: the previous implementation had the advantage of not needing to hold the entire checkpoint data for all markets in memory at once, but here we are. I imagine we will change how the checkpoints work at some point to do less allocation, so maybe it is ok for now, but something to keep in mind.

File: sequencer/src/main/kotlin/co/chainring/sequencer/core/SequencerState.kt [hunk 2]
```diff
@@ -54,10 +50,32 @@ data class SequencerState(
...
+            )
+
+            wire.read("withdrawalFees").sequence(withdrawalFees) { map, v ->
+                while (v.hasNextSequenceItem()) {
+                    val fees = WithdrawalFee.parseFrom(v.bytes())
+                    map[Symbol(fees.asset)] = fees.value.toBigInteger()
+                }
+            }
+
+            measureNanoTime {
```
Comment (bholzman) [inline sequencer/src/main/kotlin/co/chainring/sequencer/core/SequencerState.kt:77]: even though the stuff before this "should" be fast, might be worth capturing its time anyway

**Review by bholzman — APPROVED** [review]
Looks great!

File: sequencer/src/main/kotlin/co/chainring/sequencer/core/SequencerState.kt (hunk 1 above)
Comment (antonyshchenko) [reply sequencer/src/main/kotlin/co/chainring/sequencer/core/SequencerState.kt:133]: Yeah, I know. I think we could optimize this if we bypass Protobuf as serialization format for market checkpoints and write directly to the `wire`, but that is based on my assumption that `wire().write()` writes directly to the memory mapped (off-heap) file (queue). If that is the case then for each market we could stream it's checkpoint data to the queue without allocating an intermediate buffer that has to fit the entire market. I can explore this, but I propose to do it with a separate ticket.

File: sequencer/src/main/kotlin/co/chainring/sequencer/core/SequencerState.kt (hunk 2 above)
Comment (antonyshchenko) [reply sequencer/src/main/kotlin/co/chainring/sequencer/core/SequencerState.kt]: Yup, done

File: sequencer/src/main/kotlin/co/chainring/sequencer/core/SequencerState.kt (hunk 1 above)
Comment (antonyshchenko) [reply sequencer/src/main/kotlin/co/chainring/sequencer/core/SequencerState.kt:147]: Created [CHAIN-406](https://chainringinc.atlassian.net/browse/CHAIN-406) to address this

**Pull request merged**